torch/onnx/_internal/exporter.py

```diff
@@ -638,6 +641,75 @@ def model_proto(self) -> onnx.ModelProto:  # type: ignore[name-defined]
+            ...         self.register_buffer("my_buffer2", torch.tensor(4.0))
+            ...         self.conv1 = torch.nn.Conv2d(1, 32, 3, 1, bias=False)
+            ...         self.conv2 = torch.nn.Conv2d(32, 64, 3, 1, bias=False)
+            ...         self.fc1 = torch.nn.Linear(9216, 128, bias=False)
+            ...         self.fc2 = torch.nn.Linear(128, 10, bias=False)
+            ...     def forward(self, x, b):
+            ...         tensor_x = self.conv1(x)
+            ...         tensor_x = torch.nn.functional.sigmoid(tensor_x)
+            ...         tensor_x = self.conv2(tensor_x)
+            ...         tensor_x = torch.nn.functional.sigmoid(tensor_x)
+            ...         tensor_x = torch.nn.functional.max_pool2d(tensor_x, 2)
+            ...         tensor_x = torch.flatten(tensor_x, 1)
+            ...         tensor_x = self.fc1(tensor_x)
+            ...         tensor_x = torch.nn.functional.sigmoid(tensor_x)
+            ...         tensor_x = self.fc2(tensor_x)
+            ...         output = torch.nn.functional.log_softmax(tensor_x, dim=1)
+            ...         (
+            ...         self.my_buffer2.add_(1.0) + self.my_buffer1
+            ...         )  # Mutate buffer through in-place addition
+            ...         return output
+            >>> inputs = (torch.rand((64, 1, 28, 28), dtype=torch.float32), torch.randn(3))
+            >>> exported_program = torch.export.export(CustomModule(), args=inputs)
+            >>> onnx_program = torch.onnx.dynamo_export(exported_program, *inputs)
+            >>> print(onnx_program.model_signature)
+            ExportGraphSignature(
+                input_specs=[
+                    InputSpec(kind=<InputKind.PARAMETER: 2>, arg=TensorArgument(name='arg0_1'), target='conv1.weight'),
+                    InputSpec(kind=<InputKind.PARAMETER: 2>, arg=TensorArgument(name='arg1_1'), target='conv2.weight'),
+                    InputSpec(kind=<InputKind.PARAMETER: 2>, arg=TensorArgument(name='arg2_1'), target='fc1.weight'),
+                    InputSpec(kind=<InputKind.PARAMETER: 2>, arg=TensorArgument(name='arg3_1'), target='fc2.weight'),
+                    InputSpec(kind=<InputKind.BUFFER: 3>, arg=TensorArgument(name='arg4_1'), target='my_buffer2'),
+                    InputSpec(kind=<InputKind.BUFFER: 3>, arg=TensorArgument(name='arg5_1'), target='my_buffer1'),
+                    InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='arg6_1'), target=None),
+                    InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='arg7_1'), target=None)
+                ],
+                output_specs=[
+                    OutputSpec(kind=<OutputKind.BUFFER_MUTATION: 3>, arg=TensorArgument(name='add'), target='my_buffer2'),
+                    OutputSpec(kind=<OutputKind.USER_OUTPUT: 1>, arg=TensorArgument(name='_log_softmax'), target=None)
+                ]
+            )
+        """
+
+        return self._model_signature
```

Comment:
I feel this is only accurate up till the point for fx-graph in ExportedProgram. Further process inside our exporter that modifies the signature will make this outdated. Unless we apply similar io adapting to this too.

Reply:
when user model is `torch.export.ExportedProgram`, we don't have any pass that changes the signature.
The only pass applied to the resulting graph module is `fx_module = passes.InsertTypePromotion(diagnostic_context, fx_module).run()`.

Reply:
okay, this makes sense.